rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,501 @@
+
+**Decision**: Doc-comments refer to the **immediately** preceded expression. Only whitespaces are allowed in between.
+
+> Note: We are working on a generic example implementation that covers about 90% of all documentation use cases.
+
+<details>
+<summary>Arguments</summary>
+
+- (+) Doc-comments should have only one variant to reduce complexity. Referencing the next node seems straightforward. 
+- (-) A variant that references the previous node in the AST should be avoided to reduce complexity.
+- (+) Relation between documentation and the referenced implementation is straightforward and back-trackable.
+- (-) Concrete Implementation might be complex.
+- (-) Unclear if complete documentation might also need backward references. 
+    - (-), e.g., rust uses backward references only at the top of file comments.
+    - (+) Nix files can only have ONE expression; the next AST Node, in case of top-of-file comments, is thus always only that one expression. (Unlike in Rust)
+- (-) Tools need to be smart enough to understand asignments `=` and other forms of creating names for anonymous expressions. (e.g., `callPackage` and `import` )
+    - (+) Tools can still come up with other solutions that do not involve calculating everything dynamically from nix code but could also involve a static configuration.
+    - (+) The whole `tool point` is an implementation detail as long as it is not impossible. The current tool, `nixdoc`, already proves that it is possible to have static documentation to a certain degree.
+
+</details>
+
+## Single-line doc-comments (do not exist)
+
+**Observing**: Nix offers two variants of comments; single- (`#`) and multi-line comments (`/* */`). There may be use cases where it is desirable to have a form of single-line comments subtyped for doc-comment purposes. 
+
+**Considering**: Single-line comment for documentation.
+
+**Decision**: Single-line comments **cannot be used** in any form for documentation puposes.
+
+<details>
+<summary>Arguments</summary>
+
+- (-) Doc-comments should have only one variant to reduce complexity.  
+- (-) documentation will likely take up more than one line.
+- (-) If documentation grows bigger than one line, refactoring into a multiline-doc-comment must occur.
+- (+) It Would be consistent with providing variants for both nix comments.
+- (+) Offer the choice.
+- (o) Single lines can also be concatenated to form multi-line documentation.
+- (+) Takes up less vertical space
+- (-) Visually confusing when every line starts with a reserved character.
+    - (-) Potential visual conflicts with the content
+- (+) Indentation of the content is clear.
+
+</details>
+
+## Doc-comment examples
+
+The following examples demonstrate the concrete usage scenarios:
+
+`somefile.nix`
+
+~~~nix
+{
+  /**
+  Documentation for mapAttrs
+  
+  # Example
+  
+  ```
+  mapAttrs {attr = "foo"; } ...
+  ```
+  
+  # Type
+  
+  ```
+  mapAttrs :: a -> b -> c
+  ```
+  
+  */
+  mapAttrs = f: s: <...>;
```

Comment:
I initially formulated this decision because I wanted to specify only abstract placement possibilities. I was not tying that RFC to specific AST implementations.

In a concrete implementation that I'm working on. This is interpreted as follows:

-> Doc comments can appear in two places.
```nix
# (1 directly next to a value) To support anonymous expressions.
{
  foo = /** Doc */ bar: bar;
}

# (2  Above attrName the value gets assigned to) To support named expressions. (Also most common and intuitive place )
{
  /** Doc */
  foo = bar: bar;
}
```

Do you have any suggestions to formulate that abstractly?

Reply:
I'm fine with that. If you are looking for more generic options, it could as well be:

> Doc-comments refer to the **immediately** preceded expression or attribute/name expression is bound to.

In case if it is limited to only attrset, then I would keep only "attribute". If it only limited to nodes with `=`, then mentioning it like in your example might be essential.

Worth to consider also cases:

```nix
let
  /** Doc */
  foo = bar: bar;
in {
  other = foo;  # ⇐ Inherits documentation?
  inherit foo;  # ⇐ and in this case?
}
```

```nix
let
  foo = bar: bar;
in {
  /** Doc */
  inherit foo;  # ⇐ will it get attached doc?
}
```

Reply:
Yes, your first example got everything right.
The documentation is attached to the value `Lambda<bar: bar>` 

So both "other" and "foo" have "Doc" as a doc-string. 
Because both are just references to the value. 
 
But this goes into implementation details.

I can provide you that early screenshot, though:

![image](https://user-images.githubusercontent.com/50398876/235364929-951f3e74-0079-475a-ba59-0ddd1ad44737.png)


Checkout [This wip PR](https://cl.tvl.fyi/c/depot/+/8530) if you want to experiment with it,